src/Tokenizer/Tokens.php

```diff
@@ -565,25 +589,49 @@ public function getPrevNonWhitespace(int $index, ?string $whitespaces = null): ?
     }
 
     /**
-     * Get index for closest previous token of given kind.
+     * Get index for closest previous token among given kinds.
+     *
      * This method is shorthand for getTokenOfKindSibling method.
      *
-     * @param int $index token index
-     * @param list<array{int}|string|Token> $tokens        possible tokens
-     * @param bool $caseSensitive perform a case sensitive comparison
+     * @param int                                          $index         token index
+     * @param list<array{0: int, 1?: string}|string|Token> $tokens        possible tokens
+     * @param bool                                         $caseSensitive perform a case sensitive comparison
      */
     public function getPrevTokenOfKind(int $index, array $tokens = [], bool $caseSensitive = true): ?int
     {
         return $this->getTokenOfKindSibling($index, -1, $tokens, $caseSensitive);
     }
 
+    /**
+     * Get index for closest previous token of given kind.
+     *
+     * @param int                                        $index         token index
+     * @param array{0: int, 1?: string}|int|string|Token $tokens        possible tokens
+     * @param bool                                       $caseSensitive perform a case sensitive comparison
+     *
+     * @internal
+     */
+    public function tokenIndexOfKindBefore(int $index, $tokens, bool $caseSensitive = true): int
+    {
+        if (\is_int($tokens)) {
+            $tokens = [$tokens];
+        }
+        $prevIndex = $this->getPrevTokenOfKind($index, [$tokens], $caseSensitive);
+
+        if (null === $prevIndex) {
+            throw new \LogicException("No token of given kind found before index {$index}.");
```

Comment:
case 1

ref https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6680#discussion_r1026591155

we may or may not be OK to find the previous index. it's handy to assume it's there and crash otherwise. if we are OK that it is not existing, we may need to catch this error.

whatever is the reason, we want to catch the error and handle it in-code, or we bail the app and crash on console, let's use custom exception instead of most-generic one

Reply:
I'm not sure I understand. If one knows the token exists, there already is an API that returns `null` in such case. Using this new one in a `try/catch` looks less convenient, it's not its purpose.

If we leave this scenario aside, I don't think a custom exception would be relevant.

Reply:
I wonder to keep only one method ultimately.

Please consider that we may not know if there will be, or will not be, given token to be find. [eg https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/ff49030e8508ea9b71091b6fdf94de858d898cdc/src/Tokenizer/Analyzer/WhitespacesAnalyzer.php#L27 ]

Also, I would prefer to keep generic exceptions thrown by engine, while narrow, custom exceptions by app logic - better to debug it on unexpected crash.

For those 2 reasons, I would love to ask you to consider dedicated exception. If someone don't want to catch it, nothing will be worse for them. But if one would like to, it will be easier with custom exception.